Add STM32 HSEM MBOX driver and sample

celinakalus opened the pull request:
The STM32H7 dual core inter-processor communication is currently handled by an IPM driver sending interrupts via hardware semaphores; However, the new IPC service subsystem requires usage of an MBOX driver instead. Such an MBOX driver does not currently exist for processors such as the STM32H747.

This PR adds this new MBOX driver, which functions in fundamentally the same way as the STM32 HSEM IPM driver. Additionally, using this new driver, the ICMsg sample is updated to support the stm32h747i_disco board.

celinakalus committed:
drivers: mbox: Add driver for STM32 HSEM

This driver implements a simple MBOX device which supports a single
instance, two channels (one for each direction), and only signalling
mode with no data transfer. Signalling to another core is achieved by
taking and giving two hardware semaphores, similar to the STM32 HSEM
IPM driver.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

celinakalus committed:
dts: bindings: mbox: Add STM32 HSEM MBOX driver

Add a device tree binding for the new driver. Since there already
exists an IPM driver using the unsharable hardware semaphore interrupt,
the new driver is not added to any boards or SOCs per default to avoid
compatibility problems. See #37300 for the IPM driver.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

celinakalus commented:
This is my first bigger contribution to Zephyr! Please let me know if there is anything I can do better in the future.

I was not sure what to do about the hardware semaphores; Currently, I am reusing the same semaphores that the IPM driver uses, thus causing incompatibility, as only one of them can use them at the same time. My rationale was that there is probably no use case where we would need the IPM driver AND the MBOX driver at the same time. However, I can surely use two previously unused hardware semaphores for this if needed.

celinakalus commented on samples/subsys/ipc/ipc_service/icmsg/boards/stm32h747i_disco_m7.overlay:
```diff
@@ -0,0 +1,45 @@
+		reg = <0x38000000 0x08000>;
+	};
+
+	sram_rx: memory@38008000 {
```
Not the same region, one is 0x0..., the other is 0x8...

celinakalus commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,252 @@
+	ARG_UNUSED(dev);
+
+	/* Only two channels supported, one RX and one TX */
+	return 2;
```
Done.

celinakalus commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,252 @@
+
+static inline void stm32_hsem_enable_rx_interrupt(void)
+{
+	const uint32_t mask_hsem_id = (1U << MBOX_RX_HSEM_ID);
```
Done.

celinakalus commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,252 @@
+	/* Configure clock on CPU 1 */
+#if HSEM_CPU_ID == HSEM_CPU1
+	const struct mbox_stm32_hsem_conf *cfg = dev->config;
+	const struct device *const clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
```
Done (sub function).

celinakalus commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,252 @@
+
+#if HSEM_CPU_ID == HSEM_CPU1
+#define MBOX_TX_CHANNEL 0
+#define MBOX_TX_HSEM_ID CFG_HW_IPM_CPU2_SEMID
```
This sounds like an awesome idea and I would love to hear the others opinions. If this is something you decide should be done, I would gladly give implementing it a shot.

erwango replied:
A good idea, but for consistency it would require to update other usages of CFG_FOO definitions provided [here](soc/st/stm32/common/stm32_hsem.h), and in those areas, interest is lower.

celinakalus replied:
I have implemented this now; Multiple channels are now thus possible. Please let me know what you think.

celinakalus replied:
Reverted after result from another thread. Closing this here.

celinakalus commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,252 @@
+	const uint32_t mask_hsem_id = (1U << MBOX_RX_HSEM_ID);
+
+#if HSEM_CPU_ID == HSEM_CPU1
+	LL_HSEM_DisableIT_C1IER(HSEM, mask_hsem_id);
```
I did it this way because in my opinion, it has a few benefits:

1. The hsem mask is always the same for each call, so encapsulating it makes slip-ups more unlikely
2. Code is more self-documenting because the function names are now more meaningful
3. When trying to look up the function, the extra step through the defines is avoided

It is how we have been implementing this at my company for a while now. If this coding style is not desirable here, however, I will of course change it.

Depending on what is decided in [the other thread](https://github.com/zephyrproject-rtos/zephyr/pull/69680#discussion_r1509252743), argument 1 might become obsolete, however.

celinakalus replied:
It is now done like in the IPM driver.

erwango commented on samples/subsys/ipc/ipc_service/icmsg/src/main.c:
```diff
@@ -8,7 +8,12 @@
+#if defined(CONFIG_BOARD_NRF5340DK_NRF5340_CPUAPP)
 #include <hal/nrf_reset.h>
+#elif defined(CONFIG_BOARD_STM32H747I_DISCO_M7)
+#include <stm32_ll_rcc.h>
+#include <stm32h7xx_ll_system.h>
```
I might have missed an essential part here, but I don't see where these headers are required.

celinakalus replied:
You are absolutely right. There was a point at which there were some functions from there being used for testing and debugging, but I removed that by now. Will remove the headers then.

erwango reviewed (changes requested):
To be rebased on top of changes made for hwmv2 (I expect some impacts on overlays).

Anyway, thanks for the nice and welcome work so far.

celinakalus commented:
Changes required by hwmv2 have been applied.

celinakalus commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,252 @@
+	}
+
+	if (enable) {
+		stm32_hsem_clear_rx_interrupt();
```
I agree, that is problematic. I have moved the clearing over to the initialization. Correct me if I am wrong, but since we have two separate registers for the interrupt active bit (one for each core), we need to clean it from both cores. I have implemented it this way.

celinakalus replied:
Resolving this now.

arnopo commented on samples/subsys/ipc/ipc_service/icmsg/remote/boards/stm32h747i_disco.overlay:
```diff
@@ -0,0 +1,47 @@
+		status = "okay";
+		interrupts = < 0x7e 0x0 >;
+		st,hsem_ids_tx = <11>;
+		st,hsem_ids_rx = <10>;
```
If we decide to declare list of used channels, I would only list here the ID supported without the RX/TX  property, I look like a duplication of properties defined  in the ipc0 node

what about " st, hsem-channels = <10>, <11>; ?

celinakalus replied:
A single hardware semaphore cannot be bidirectional, so the user would have to make sure not to use a single MBOX channel for both sending and receiving. Is that a problem?

celinakalus replied:
This is now obsolete, since these properties are removed.

arnopo commented on samples/subsys/ipc/ipc_service/icmsg/remote/boards/stm32h747i_disco.overlay:
```diff
@@ -0,0 +1,47 @@
+		#mbox-cells = <1>;
+	};
+
+	ipc0: ipc0 {
```
seems also that this could be in DT board, and juste enbale the node here

erwango replied:
Nodes in boards should be enabled (or actually they are systematically today for reasons, I think this should be kept that way).

celinakalus replied:
I would not put this into the board, actually, since it requires splitting the `sram4` memory into two regions, and other applications might depend on that region -- most notably the IPM driver for the same board.

arnopo reviewed (commented):
I have a few comments, but the main point is how to ensure that we use the expected channels for the mbox.
In the end, I just wonder if the node should not be added in the Soc and Board DT with channel declaration. That would ensure that the default channel usage is coherent with the ones defined in https://elixir.bootlin.com/zephyr/latest/source/soc/arm/st_stm32/common/stm32_hsem.h.

celinakalus commented on samples/subsys/ipc/ipc_service/icmsg/remote/boards/stm32h747i_disco.overlay:
```diff
@@ -0,0 +1,47 @@
+	};
+};
+
+&pinctrl {
```
Same place where the (IPM) mailbox is defined

erwango replied:
ipm mailbox is defined under soc:
https://github.com/zephyrproject-rtos/zephyr/blob/cbef9169733399d80f6e5057faf5b95490f9644f/dts/arm/st/h7/stm32h7_dualcore.dtsi#L18-L20

celinakalus replied:
Ahh, I was wrong then.

celinakalus commented on samples/subsys/ipc/ipc_service/icmsg/remote/boards/stm32h747i_disco.overlay:
```diff
@@ -0,0 +1,47 @@
+};
+
+&pinctrl {
+	mbox: mbox@58026400 {
```
I was cautious of doing that because of the already existing IPM mailbox which is already used and is in conflict with this MBOX (if one uses the same HSEMs). I would be thankful for any hints how to proceed in a manner that preserves compatibility with existing code.

celinakalus replied:
These DT nodes have now been merged together and the MBOX is defined in the device tree of the board instead of the overlay.

celinakalus commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,257 @@
+#define MAX_RX_CHANNELS DT_INST_PROP_LEN(0, st_hsem_ids_rx)
+#define MAX_CHANNELS (MAX_RX_CHANNELS + MAX_TX_CHANNELS)
+
+
```
I've run checkpatch and it didn't catch that, weird. Will fix.

arnopo replied:
> I've run checkpatch and it didn't catch that, weird. Will fix.

try with the --strict option.

arnopo replied:
double blank line still there

celinakalus commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,257 @@
+		.bus = DT_INST_CLOCKS_CELL(0, bus),
+		.enr = DT_INST_CLOCKS_CELL(0, bits)
+	},
+};
```
Done.

erwango commented on samples/subsys/ipc/ipc_service/icmsg/boards/stm32h747i_disco.overlay:
```diff
@@ -0,0 +1,45 @@
+		clocks = < &rcc 0xe0 0x2000000 >;
+		status = "okay";
+		interrupts = < 0x7d 0x0 >;
+		#mbox-cells = <1>;
```
Since this is the same hw, I think we should use only one device tree description.
The SW implementation choice should be decided by Kconfig symbols.

In your case, it would mean that you would reuse the node and populate it with the new required properties which would be left unused by the legacy IPM driver.

erwango replied:
Note on this comment:
It does not prevent the Kconfig driver selection to depends on dts for instance with `DT_HAS_ZEPHYR_IPC_MSG_ENABLED`

erwango commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,257 @@
+{
+	const struct mbox_stm32_hsem_conf *cfg = dev->config;
+
+	return (cfg->rx_ch_mask & BIT(ch));
```
After discussion with @arnopo :
 Go with "Implement a test using [CFG_HW_IPM_CPU1_SEMID](https://elixir.bootlin.com/zephyr/latest/C/ident/CFG_HW_IPM_CPU1_SEMID) or [CFG_HW_IPM_CPU2_SEMID](https://elixir.bootlin.com/zephyr/latest/C/ident/CFG_HW_IPM_CPU2_SEMID), depending on the HSEM_CPU_ID":
- remove properties:  st,hsem_ids_tx/rx
- Implement check vs header in current function.

Ideally, we could make it more simple by moving definitions from stm32_hsem.h to actual device tree headers that could be included in dt and driver and remove the need for a check. Not requested today since this would require additional unrelated work (split headers, ..).

celinakalus replied:
I am not sure how to understand this. Do I
1. allow ONLY the same IPM semaphores, OR
2. allow arbitrary semaphore IDs ABOVE those IPM semaphores?

celinakalus replied:
@erwango @arnopo I hope you are doing well. Could you give me some additional clarification about how I should be implementing this? See my previous comment. Thank you in advance!

erwango replied:
I'd say allow only the same IPM semaphores

celinakalus committed:
samples: ipc: icmsg: Add received bytes metric

By adding this metric, any problem with the remote core sending messages
is observable.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

celinakalus committed:
samples: ipc: icmsg: Do not return busy error codes

When sending a lot of data, a busy error code returned by the IPC is to
be expected. But if the last attempted send returns busy, this error
code is returned to the calling function, causing an error log message.
Fix this by setting return variable to 0 on this acceptable error.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

celinakalus committed:
samples: ipc: icmsg: Add support for stm32h747i_disco

With the addition of the STM32 hardware semaphore MBOX driver, ICMsg is
now supported on the stm32h747i_disco board.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

celinakalus committed:
samples: drivers: mbox: Add support for stm32h747i_disco board

With the addition of the STM32 hardware semaphore MBOX driver, this
sample is now supported by the board.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

celinakalus committed:
dts: stm32h7_dualcore: Add MBOX driver

Adding the new STM32 hardware semaphore driver into the device tree.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

celinakalus committed:
dts: bindings: ipm: Add dummy mbox-cells property

The MBOX driver interface expects a device tree property '#mbox-cells'
which is not known by the IPM driver. This causes build problems when
both drivers are given for a single shared DT node.

To fix this problem for this driver specifically, add a dummy
'#mbox-cells' property to the bindings of the STM32 HSEM IPM driver.
This does not affect any other IPM driver, and the STM32 HSEM IPM driver
is still functioning with this dummy property.

Signed-off-by: Celina Sophie Kalus <hello@celinakalus.de>

erwango commented:
@celinakalus I can see that you made some changes. Is that the final version including all required changes ?

celinakalus commented:
Yes, the version now should be final, with all requested changes applied, unless I stated otherwise in a comment. @erwango

erwango reviewed (dismissed):
ok for me but review from @arnopo or @carlocaione is required.

arnopo commented on drivers/mbox/mbox_stm32_hsem.c:
```diff
@@ -0,0 +1,261 @@
+#define LOG_LEVEL CONFIG_MBOX_LOG_LEVEL
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+LOG_MODULE_REGISTER(mbox_stm32_hsem_ipc);
```
#include  should be defined  in alphabetical order and you could use LOG_MODULE_REGISTER

```c
#include <zephyr/device.h>
#include <zephyr/drivers/clock_control.h>
#include <zephyr/drivers/clock_control/stm32_clock_control.h>
#include <zephyr/drivers/mbox.h>
#include <zephyr/irq.h>
#include <zephyr/logging/log.h>


#include "stm32_hsem.h"

LOG_MODULE_REGISTER(mbox_stm32_hsem_ipc, CONFIG_MBOX_LOG_LEVEL);
```

celinakalus replied:
Thank you; I see I've been looking at the IPM driver too long to realize what should be done better. Fixed it with force push.

arnopo reviewed (dismissed):
2 minor comments else LGTM

celinakalus commented:
Force push to address comments by @arnopo; No code changed, tests should pass soon.
